### DISCUSSION TITLE
Strategies for per-install reports (clients)

### DISCUSSION THREAD

**Discussion by Zidane** [QuestPDF/QuestPDF#804] [Q&A] [answered]
Greetings QuestPDF Community!

Unsure if this is the right place for this but looking for ideas here. I am building an application that very likely will have several clients who will all want their reports to look a specific way tailored to their needs. The data set for the report will be unchanged but the document layout will obviously change. Given that QuestPDF is very much a code-first approach, has anyone else used this library in this manner and has ideas here? Ideally, I need a way to retain RDLC's modularity here with being able to keep those reports outside of the compiled binaries. It was never a fun experience but at least an end user could open up a RDLC and make changes.

One approach I considered was compiling DLLs per-client with just their reports. Which makes me feel icky. Another approach was to do the document layouts in xaml first, write to file, and then have QuestPDF load that xaml as the Document. Which iirc it can do this.

Any other thoughts? I'd appreciate any input here. Loving the library thus far!

**Comment by MarcinZiabek** [comment, marked as answer]
Thank you for the kind words, and sorry for the late reply!

You don't need a separate C# file per company. Because QuestPDF documents are plain C# code, there are two levels of flexibility, depending on how different the invoices need to be:

1) Same layout, different branding (most common case). Create a single invoice document class and pass a per-company "theme" object (colors, logo, fonts, spacing) alongside the invoice data. Every style call accepts runtime values, so branding becomes just another input. The [Components](https://www.questpdf.com/concepts/code-patterns/components.html) and [Content styling](https://www.questpdf.com/concepts/code-patterns/content-styling.html) patterns show how to structure this cleanly, and [Conditional formatting](https://www.questpdf.com/concepts/dynamic-composition/conditional-formatting.html) shows how to apply styles based on values or settings.

2) Structurally different layouts. When companies need genuinely different document structures, you can drive the layout itself from data or configuration — for example, define which sections appear and in what order, or which table columns are shown, per company:

[Mixed content blocks](https://www.questpdf.com/concepts/dynamic-composition/mixed-content-blocks.html) — model the document as a runtime-defined sequence of blocks,
[Configurable tables](https://www.questpdf.com/concepts/dynamic-composition/configurable-tables.html) — define table columns from configuration objects,
[Nested content](https://www.questpdf.com/concepts/dynamic-composition/nested-content.html) — render hierarchical data of arbitrary depth.

In practice, most teams combine both: shared components for common elements (line items, totals, footer) plus a small amount of per-company composition. If one customer truly needs a bespoke design, a dedicated class for that one case is perfectly fine too — it's ordinary C#, so standard techniques like composition and inheritance all apply.

I'm closing this discussion as answered, but feel free to follow up if anything is unclear. Good luck with your project!
